File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int n = height.length ;
        int left = 0 , right = n -1 ;
        int leftMax = height[left] ;
        int rightMax = height[right] ;
        int result = 0 ;
        while(left < right){
            if(height[left] <= height[right]){
                if(height[left] > leftMax) leftMax = height[left] ;
                result += leftMax - height[left] ;
                left++ ;

            }
            else {
                if(height[right] > rightMax ) rightMax = height[right] ;
                result += rightMax - height[right] ;
                right-- ;
            }
        }
        return result ;
        
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        l , r = 0 , len(height) - 1
        total = 0
        leftMax , rightMax = height[l] , height[r] 
        while l <= r :
            if leftMax <= rightMax :
                leftMax = max(leftMax,height[l])
                total += leftMax - height[l] 
                l+=1
            else :
                rightMax = max(rightMax,height[r])
                total += rightMax - height[r] 
                r-=1
        return total

        